tests/model_selection/test_hyperband.py

```diff
@@ -0,0 +1,431 @@
+            # split and we're testing on the entire train dataset, not only the
+            # validation/test set.
+            assert abs(score - search.best_score_) < 0.1
+
+        assert type(search.best_estimator_) == type(model)
+        assert isinstance(search.best_params_, dict)
+
+        num_fit_models = len(set(search.cv_results_["model_id"]))
+        num_pf_calls = sum(
+            [v[-1]["partial_fit_calls"] for v in search.model_history_.values()]
+        )
+        models = {9: 17, 15: 17, 20: 17, 27: 49, 30: 49, 81: 143}
+        pf_calls = {9: 69, 15: 101, 20: 144, 27: 357, 30: 379, 81: 1581}
+        assert num_fit_models == models[max_iter]
+        assert num_pf_calls == pf_calls[max_iter]
+
+        best_idx = search.best_index_
+        if isinstance(model, ConstantFunction):
+            assert search.cv_results_["test_score"][best_idx] == max(
+                search.cv_results_["test_score"]
+            )
+        model_ids = {h["model_id"] for h in search.history_}
+
+        if math.log(max_iter, 3) % 1.0 == 0:
+            # log(max_iter, 3) % 1.0 == 0 is the good case when max_iter is a
+            # power of search.aggressiveness
+            # In this case, assert that more models are tried then the max_iter
+            assert len(model_ids) > max_iter
+        else:
+            # Otherwise, give some padding "almost as many estimators are tried
+            # as max_iter". 3 is a fudge number chosen to be the minimum; when
+            # max_iter=20, len(model_ids) == 17.
+            assert len(model_ids) + 3 >= max_iter
+        assert all("bracket" in id_ for id_ in model_ids)
+
+    _test_basic()
+
+
+@pytest.mark.parametrize("max_iter,aggressiveness", [(27, 3), (30, 4)])
+def test_hyperband_mirrors_paper_and_metadata(max_iter, aggressiveness):
+    @gen_cluster(client=True, timeout=5000)
+    def _test_mirrors_paper(c, s, a, b):
+        X, y = make_classification(n_samples=10, n_features=4, chunks=10)
+        model = ConstantFunction()
+        params = {"value": np.random.rand(max_iter)}
+        alg = HyperbandSearchCV(
+            model,
+            params,
+            max_iter=max_iter,
+            random_state=0,
+            aggressiveness=aggressiveness,
+        )
+        yield alg.fit(X, y)
+        assert isinstance(alg.metadata["brackets"], list)
+        assert isinstance(alg.metadata_["brackets"], list)
+        assert set(alg.metadata.keys()) == {
+            "estimators",
+            "partial_fit_calls",
+            "brackets",
+        }
+        assert all(
+            set(v.keys())
+            == {
+                "bracket",
+                "estimators",
+                "partial_fit_calls",
+                "SuccessiveHalvingSearchCV params",
+                "decisions",
+            }
+            for bracket_info in [alg.metadata_["brackets"], alg.metadata["brackets"]]
+            for v in bracket_info
+        )
```

Comment:
This comprehension is too complication for a comprehension. Expand it. Also, test to make sure that the values are the same.

Reply:
I don't see why this test doesn't say `alg.metadata == alg.metadata()`. Patience isn't used at all.